url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/WebSocketImpl.java

```diff
@@ -0,0 +1,737 @@
+			}
+		}
+		return false;
+	}
+
+	private void decodeFrames( ByteBuffer socketBuffer ) {
+
+		List<Framedata> frames;
+		try {
+			frames = draft.translateFrame( socketBuffer );
+			for( Framedata f : frames ) {
+				if( DEBUG )
+					System.out.println( "matched frame: " + f );
+				Opcode curop = f.getOpcode();
+				boolean fin = f.isFin();
+
+				if( curop == Opcode.CLOSING ) {
+					int code = CloseFrame.NOCODE;
+					String reason = "";
+					if( f instanceof CloseFrame ) {
+						CloseFrame cf = (CloseFrame) f;
+						code = cf.getCloseCode();
+						reason = cf.getMessage();
+					}
+					if( readystate == READYSTATE.CLOSING ) {
+						// complete the close handshake by disconnecting
+						closeConnection( code, reason, true );
+					} else {
+						// echo close handshake
+						if( draft.getCloseHandshakeType() == CloseHandshakeType.TWOWAY )
+							close( code, reason, true );
+						else
+							flushAndClose( code, reason, false );
+					}
+					continue;
+				} else if( curop == Opcode.PING ) {
+					wsl.onWebsocketPing( this, f );
+					continue;
+				} else if( curop == Opcode.PONG ) {
+					wsl.onWebsocketPong( this, f );
+					continue;
+				} else if( !fin || curop == Opcode.CONTINUOUS ) {
+					if( curop != Opcode.CONTINUOUS ) {
+						if( current_continuous_frame_opcode != null )
+							throw new InvalidDataException( CloseFrame.PROTOCOL_ERROR, "Previous continuous frame sequence not completed." );
+						current_continuous_frame_opcode = curop;
+					} else if( fin ) {
+						if( current_continuous_frame_opcode == null )
+							throw new InvalidDataException( CloseFrame.PROTOCOL_ERROR, "Continuous frame sequence was not started." );
+						current_continuous_frame_opcode = null;
+					} else if( current_continuous_frame_opcode == null ) {
+						throw new InvalidDataException( CloseFrame.PROTOCOL_ERROR, "Continuous frame sequence was not started." );
+					}
+					try {
+						wsl.onWebsocketMessageFragment( this, f );
+					} catch ( RuntimeException e ) {
+						wsl.onWebsocketError( this, e );
+					}
+
+				} else if( current_continuous_frame_opcode != null ) {
+					throw new InvalidDataException( CloseFrame.PROTOCOL_ERROR, "Continuous frame sequence not completed." );
+				} else if( curop == Opcode.TEXT ) {
+					try {
+						wsl.onWebsocketMessage( this, Charsetfunctions.stringUtf8( f.getPayloadData() ) );
+					} catch ( RuntimeException e ) {
+						wsl.onWebsocketError( this, e );
+					}
+				} else if( curop == Opcode.BINARY ) {
+					try {
+						wsl.onWebsocketMessage( this, f.getPayloadData() );
+					} catch ( RuntimeException e ) {
+						wsl.onWebsocketError( this, e );
+					}
+				} else {
+					throw new InvalidDataException( CloseFrame.PROTOCOL_ERROR, "non control or continious frame expected" );
+				}
+			}
+		} catch ( InvalidDataException e1 ) {
+			wsl.onWebsocketError( this, e1 );
+			close( e1 );
+			return;
+		}
+	}
+
+	private void close( int code, String message, boolean remote ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.